patches/server/0934-Add-getDrops-to-BlockState-and-isPreferredTool-to-Bl.patch

```diff
@@ -0,0 +1,100 @@
++import org.bukkit.entity.Entity;
++import org.bukkit.inventory.ItemStack;
+ import org.bukkit.material.Attachable;
+ import org.bukkit.material.MaterialData;
+ import org.bukkit.metadata.MetadataValue;
+@@ -330,5 +338,32 @@ public class CraftBlockState implements BlockState {
+     public boolean isCollidable() {
+         return this.data.getBlock().hasCollision;
+     }
++
++    @Override
++    public Collection<ItemStack> getDrops() {
++        return this.getDrops(null);
++    }
++
++    @Override
++    public Collection<ItemStack> getDrops(ItemStack item) {
++        return this.getDrops(item, null);
++    }
++
++    @Override
++    public Collection<ItemStack> getDrops(ItemStack item, Entity entity) {
++        net.minecraft.world.item.ItemStack nms = CraftItemStack.asNMSCopy(item);
++
++        // Modelled off EntityHuman#hasBlock
++        if (item == null || !data.requiresCorrectToolForDrops() || nms.isCorrectToolForDrops(data)) {
++            return net.minecraft.world.level.block.Block.getDrops(
++                data,
++                world.getHandle(),
++                position,
++                world.getHandle().getBlockEntity(position), entity == null ? null : ((CraftEntity) entity).getHandle(), nms
++            ).stream().map(CraftItemStack::asBukkitCopy).collect(Collectors.toList());
++        } else {
++            return Collections.emptyList();
++        }
++    }
+     // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+index e4c15fcbd21f70836c26133ef10f3d0da9b6c238..2b388e83110f45d437debcc4d4db653c7292326a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
++++ b/src/main/java/org/bukkit/craftbukkit/block/data/CraftBlockData.java
+@@ -33,7 +33,9 @@ import org.bukkit.craftbukkit.CraftSoundGroup;
+ import org.bukkit.craftbukkit.CraftWorld;
+ import org.bukkit.craftbukkit.block.CraftBlock;
+ import org.bukkit.craftbukkit.block.CraftBlockSupport;
++import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
++import org.bukkit.inventory.ItemStack;
+ 
+ public class CraftBlockData implements BlockData {
+ 
+@@ -620,4 +622,16 @@ public class CraftBlockData implements BlockData {
+         return this.state.isRandomlyTicking();
+     }
+     // Paper end
++
++    // Paper start - Preferred tool API
++    @Override
++    public boolean isPreferredTool(org.bukkit.inventory.ItemStack item) {
++        net.minecraft.world.item.ItemStack nms = CraftItemStack.asNMSCopy(item);
++        return isPreferredTool(state, nms);
++    }
++
++    private boolean isPreferredTool(net.minecraft.world.level.block.state.BlockState iblockdata, net.minecraft.world.item.ItemStack nmsItem) {
++        return !iblockdata.requiresCorrectToolForDrops() || nmsItem.isCorrectToolForDrops(iblockdata);
++    }
```

Comment:
You don't need a private method here, just put this in the isPreferredTool implementation.